src/targets/gpu/kernels/include/migraphx/kernels/tuple.hpp

```diff
@@ -0,0 +1,164 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2015-2024 Advanced Micro Devices, Inc. All rights reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+#ifndef MIGRAPHX_GUARD_KERNELS_TUPLE_HPP
+#define MIGRAPHX_GUARD_KERNELS_TUPLE_HPP
+
+#include <migraphx/kernels/array.hpp>
+
+namespace migraphx {
+
+namespace tuple_detail {
+
+template <class T, index_int N>
+struct element_storage
+{
+    /* [[no_unique_address]] */ T element;
```

Comment:
What do you mean by this comment?

Reply:
This attribute is part of C++20, I dont know if we need to enable C++20 or if there is another attribute. Let me look into that.

This attribute is to make empty elements in the tuple to not increase the size of the tuple. So `tuple<empty_type, int> == sizeof(int)`. To do this the compiler needs to assign an address to `element` that is shared with other elements(it doesnt matter that the address is shared since an empty has no data to be read).

C++ requires each member variable to have a unique address, so `[[no_unique_address]]` tells the compiler to let it have a non-unique address that is an address that can be shared with other member variables. There are ways to achieve this without using this attribute through inheritance but this introduces other complexities to deal with(such as ensuring that `tuple<tuple<empty>, tuple<empty>>` is not ambiguous).

Reply:
It works fine without problems so I uncommented it.

Reply:
Understood. I think it would get ignored though if its not supported whether w have C++20 on or not though wouldn't it? Doesn't hurt to have it in then if it simplifies things.

Reply:
It says in clang docs that its supported in C++11 or later:

https://releases.llvm.org/15.0.0/tools/clang/docs/AttributeReference.html#no-unique-address

So it should be fine.

Reply:
Yeah I'd say leave it in then